source/Halibut/Diagnostics/LogEventStorage.cs

```diff
@@ -0,0 +1,45 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+
+namespace Halibut.Diagnostics
+{
+    public class LogEventStorage
+    {
+        static readonly LogEvent[] EmptyLogs = new LogEvent[0];
+
+        readonly ConcurrentDictionary<string, Queue<LogEvent>> events = new ConcurrentDictionary<string, Queue<LogEvent>>();
+        readonly Queue<string> lastLogEndpoint = new Queue<string>();
+
+        public LogEventStorage()
+        {
+        }
+
+        public IList<LogEvent> GetLogs(string endpoint)
+        {
+            if (events.TryGetValue(endpoint, out var logs))
+                return logs.ToArray();
+
+            return EmptyLogs;
+        }
+
+        public void AddLog(string endpoint, LogEvent logEvent)
+        {
+            lock (events)
```

Comment:
Logs would be inserted often, but retrieved rarely. I would optimise this for the write scenario, which would also make it a lot simpler. Just keep a single queue of all messages `ConcurrentQueue` and filter it when logs are requests. No locks are needed then.

Reply:
On second thoughts, that does allow a single endpoint to monopolise the messages (as it also does with this new code). Previously a single endpoint could only have 100 messages. perhaps also add that logic...